### PR TITLE
Add thread parameter to solveequihash benchmark

### LINKED ISSUE #1147: Make "zcbenchmark solveequihash" take a number-of-threads argument
Currently the benchmark is always single-threaded. Multi-threaded mining performance can't be inferred from single-threaded performance, since memory bandwidth shared between cores is a significant constraint.

Note that this is about benchmarking what the effect of genproclimit would be, not making the solver internally multithreaded.

### PR DESCRIPTION
**Pull request by str4d**
Closes #1147

### CODE REVIEW

**Comment by str4d** [comment]
Buildbot will need to be updated to run this as a new benchmark, which will require extending `PerformanceTestRunner` to accept arguments for benchmarks. Then the call to run the multithreaded benchmark will be something like:

```
time('solveequihash', [numThreads], name='solveequihash-%dthreads' % numThreads)
```

**Comment by str4d** [comment]
Proposed Buildbot config changes are in https://github.com/Electric-Coin-Company/bbotzc/pull/24

**Comment by bitcartel** [comment]
ACK

File: qa/zcash/performance-measurements.sh [hunk 1]
```diff
@@ -77,10 +77,15 @@ case "$1" in
...
+                zcash_rpc zcbenchmark verifyjoinsplit 1000 "\"$RAWJOINSPLIT\""
                 ;;
             solveequihash)
-                zcash_rpc zcbenchmark solveequihash 10
+                if [ "$#" -ge "3" ]
+                then
+                    zcash_rpc zcbenchmark solveequihash 10 "$3"
+                else
+                    zcash_rpc zcbenchmark solveequihash 10
+                fi
```
Comment (daira) [inline qa/zcash/performance-measurements.sh]: See http://stackoverflow.com/questions/9057387/process-all-arguments-except-the-first-one for a simpler way to do this.

File: src/zcbenchmarks.cpp [hunk 2]
```diff
@@ -116,10 +117,24 @@ double benchmark_solve_equihash()
                                     nonce.begin(),
                                     nonce.size());
 
-    timer_start();
+    if (time)
+        timer_start();
```
Comment (daira) [inline src/zcbenchmarks.cpp]: Why is this change needed?

**Comment by daira** [comment]
ACK, my comments are not blockers.

File: src/zcbenchmarks.cpp (hunk 2 above)
Comment (str4d) [inline src/zcbenchmarks.cpp]: `timer_start()` changes a file-global variable, so the threads would interfere with the outer timing. Also, we don't need the (admittedly minimal) overhead of the function call inside each thread.

**Comment by str4d** [comment]
@zkbot r+

File: src/zcbenchmarks.cpp (hunk 2 above)
Comment (daira) [inline src/zcbenchmarks.cpp]: Ah I see now. Note that the threaded benchmark includes the time to serialise and hash the block header+nonce, while the unthreaded benchmark doesn't. I don't think this is a big problem, but probably it's more realistic as well as more consistent to include this always.

**Comment by Vorksholk** [comment]
Since this benchmark runs solvers on x threads but waits until all threads have finished before starting the next batch of solvers, it doesn't seem to be terribly representative of real-world throughput, since towards the end of the benchmark computational resources are sitting idle.

For example, running the zcbenchmark on 6 threads for two cycles performs the first cycle, and doesn't start the 2nd cycle until all threads of the first cycle have finished. On my 6-core system, I end up waiting ~10-15 seconds for the final thread to finish its work. 

A 'simple' fix would be modifying the benchmark to solve a "pool" of x equihashes and allow threads to draw from the pool of equihashes on-demand, so that longer benchmarks would come closer and closer to approximating real-world performance since the only idle resources would exist when the pool is empty and some threads have finished their work.
